### PR TITLE
ci: Auto-generate and publish release notes for minor releases

### PR DESCRIPTION
**Pull request by tylerbutler**
The transform:releaseNotes command was added in #22466 to enable automated release note creation for minor releases. This change updates the CI pipeline to take advantage of the new capabilities.

[AB#19242](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/19242)

### CODE REVIEW

File: .github/workflows/push-tag-create-release.yml [hunk 1]
```diff
@@ -116,11 +118,16 @@ jobs:
...
-        if: ${{ !(fromJson(env.RELEASE_JSON).releaseType == 'patch' && fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'client') && fromJson(env.RELEASE_JSON).packageOrReleaseGroup != 'build-tools' }}
+
+      - name: Generate minor/major release notes
+        # This content is the "default" release notes - this should be used for minor and major releases of client and
+        # server release groups. To use this, a release group needs to be using changesets. Build-tools does not use
+        # changesets, so it is excluded.
+        if: fromJson(env.RELEASE_JSON).releaseType != 'patch' && (fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'client' || fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'server')
         run: |
-          echo "This is a **${{ fromJson(env.RELEASE_JSON).releaseType }}** release." > auto-changelog.md
+          flub publish releaseNotes \
```
Comment (tylerbutler) [inline .github/workflows/push-tag-create-release.yml:128]:
```suggestion
          flub transform releaseNotes \
```

File: .github/workflows/push-tag-create-release.yml [hunk 2]
```diff
@@ -116,11 +118,16 @@ jobs:
...
+
+      - name: Generate minor/major release notes
+        # This content is the "default" release notes - this should be used for minor and major releases of client and
+        # server release groups. To use this, a release group needs to be using changesets. Build-tools does not use
+        # changesets, so it is excluded.
+        if: fromJson(env.RELEASE_JSON).releaseType != 'patch' && (fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'client' || fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'server')
         run: |
-          echo "This is a **${{ fromJson(env.RELEASE_JSON).releaseType }}** release." > auto-changelog.md
+          flub transform releaseNotes \
+          --inFile RELEASE_NOTES/${{ fromJson(env.RELEASE_JSON).version }}.md
```
Comment (tylerbutler) [inline .github/workflows/push-tag-create-release.yml:129]:
```suggestion
          --inFile RELEASE_NOTES/${{ fromJson(env.RELEASE_JSON).version }}.md \
```

File: .github/workflows/push-tag-create-release.yml [hunk 3]
```diff
@@ -116,11 +118,16 @@ jobs:
...
+      - name: Generate minor/major release notes
+        # This content is the "default" release notes - this should be used for minor and major releases of client and
+        # server release groups. To use this, a release group needs to be using changesets. Build-tools does not use
+        # changesets, so it is excluded.
+        if: fromJson(env.RELEASE_JSON).releaseType != 'patch' && (fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'client' || fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'server')
         run: |
-          echo "This is a **${{ fromJson(env.RELEASE_JSON).releaseType }}** release." > auto-changelog.md
+          flub transform releaseNotes \
+          --inFile RELEASE_NOTES/${{ fromJson(env.RELEASE_JSON).version }}.md \
+          --outFile auto-changelog.md
```
Comment (alexvy86) [inline .github/workflows/push-tag-create-release.yml:130]: `release-notes.md`? I see changelog and think of the individual package changelogs.

File: .github/workflows/push-tag-create-release.yml [hunk 4]
```diff
@@ -116,11 +118,16 @@ jobs:
           --tag-prefix ${{ fromJson(env.RELEASE_JSON).packageOrReleaseGroup }}_v \
           --output auto-changelog.md \
           --template .github/workflows/data/patch-changelog.hbs
-      - name: Generate changelog
-        # This changelog is a basically empty one used for everything except client patches and build-tools.
-        if: ${{ !(fromJson(env.RELEASE_JSON).releaseType == 'patch' && fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'client') && fromJson(env.RELEASE_JSON).packageOrReleaseGroup != 'build-tools' }}
+
+      - name: Generate minor/major release notes
```
Comment (alexvy86) [inline .github/workflows/push-tag-create-release.yml:122]: Line 151 refers to `the "Generate changelog"` step, now it probably wants to read `created by the steps above` or something along those lines.

File: .github/workflows/push-tag-create-release.yml [hunk 5]
```diff
@@ -101,10 +103,10 @@ jobs:
...
           retention-days: 7
 
-      # Generate changelog
-      - name: Generate changelog
-        # This changelog is only for client patch releases and build-tools releases
-        if: (fromJson(env.RELEASE_JSON).releaseType == 'patch' && fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'client') || fromJson(env.RELEASE_JSON).packageOrReleaseGroup == 'build-tools'
+      # Generate release notes
+      - name: Generate patch release notes
+        # This content is only for patch releases
+        if: fromJson(env.RELEASE_JSON).releaseType == 'patch'
```
Comment (alexvy86) [inline .github/workflows/push-tag-create-release.yml:109]: Will this now run for patch releases of anything including Independent packages and server? Also, build-tools used to get this step for every kind of release, now it'll be only for its patch releases, and we're explicitly excluding it from the next step, so there won't be auto-generation of any kind for build-tools minors and majors?

File: .github/workflows/push-tag-create-release.yml (hunk 3 above)
Comment (tylerbutler) [reply .github/workflows/push-tag-create-release.yml:130]: I'm using this name because its the same name that the patch releases generate.

Comment (tylerbutler) [reply .github/workflows/push-tag-create-release.yml:130]: I guess I can change both of them though.

Comment (alexvy86) [reply .github/workflows/push-tag-create-release.yml:130]: Yeah, that's what I had in mind :)

File: .github/workflows/push-tag-create-release.yml (hunk 5 above)
Comment (tylerbutler) [reply .github/workflows/push-tag-create-release.yml:109]: > Will this now run for patch releases of anything including Independent packages and server?

Yes. However, I updated it to leave every release except client/server as draft.

> Also, build-tools used to get this step for every kind of release, now it'll be only for its patch releases, and we're
> explicitly excluding it from the next step, so there won't be auto-generation of any kind for build-tools minors and
> majors?

The workflow will fail if the release notes are missing, so I was just going to let them fail. But instead I inverted
the logic from the major/minor releases in this one, so now one of the two steps will be executed on each workflow run.

**Review by alexvy86 — APPROVED** [review]
Last nit: the comment on line 117 could go away now, or be copied/hoisted to line 106. But not critical.

**Pull request merged**